dll/win32/winmm/playsound.c

```diff
@@ -92,7 +92,21 @@ static HMMIO	get_mmioFromProfile(UINT uFlags, LPCWSTR lpszName)
      *      HKCU\AppEvents\Schemes\Apps\.Default
      *      HKCU\AppEvents\Schemes\Apps\<AppName>
      */
-    if (RegOpenKeyW(HKEY_CURRENT_USER, wszKey, &hRegSnd) != 0) goto none;
+    err = RegOpenCurrentUser(KEY_READ, &hUserKey);
```

Comment:
For information, the reason why `RegOpenKey(Ex)(HKEY_CURRENT_USER, ...` is not completely equivalent to `RegOpenCurrentUser`, is because in the case of `RegOpenKey`, there is a caching being done for the "predefined" handles (like HKEY_CURRENT_USER), and when impersonating, the cached value is the one still used, instead of a new one corresponding to the impersonated user.

You can add this command ^ above in your commit log, this will help understand what's going on.

Reply:
Another HKEY_CURRENT_USER that may be buggy for the same reasons, in winmm, is this one:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/winmm/midimap/midimap.c;hb=83442919e4df0067a7bc72f351fefed667c1737a#l207

Reply:
Thanks for the info, I found that Windows has two undocumented functions: `WinmmLogon` and `WinmmLogoff`

winlogon.exe from WinXP SP3:
![WinXP_WinMM](https://user-images.githubusercontent.com/86486978/186997042-efefa84a-0978-4e24-a099-ed924ac76888.PNG)

`WinmmLogon` maybe the way Windows gets user token directly from winlogon and use it to access the appropriate key.

Reply:
May also be related to managing the audio service (ensuring it's started)?

Reply:
Based on my research, `WinmmLogon` is used to initialize the audio system on Windows. This function doesn't take user token as a parameter.

I think implementing it is not necessary. A better solution maybe to implement ROS-specific functions in winmm and dynamically load them from winlogon. This way we can use Wine's code with least modification.

Sources:
[https://forum.thegamecreators.com/thread/176905](https://forum.thegamecreators.com/thread/176905)

[https://www.writebug.com/git/greenlight/win32-asm-calculator/src/branch/master/src/Calculator(源程序)/Calculator/Calculator/winmm.inc](https://www.writebug.com/git/greenlight/win32-asm-calculator/src/branch/master/src/Calculator(源程序)/Calculator/Calculator/winmm.inc)

[https://rsdn.org/forum/info/FAQ.mmedia.winmminit](https://rsdn.org/forum/info/FAQ.mmedia.winmminit)

Reply:
The better solution may be to implement ROS-specific functions in winmm and dynamically load them from winlogon, for example:

```
BOOL WINAPI WinmmRosLogon(_In_ HANDLE hUserToken);
VOID WINAPI WinmmRosLogoff(VOID);
```

Reply:
But then, why adding those functions at all to begin with? How is this going to help with having PlaySound reading the correct user data from the registry?

Reply:
`WinmmRosLogon` will be the way winmm gets user token directly from the calling process and will call `RegOpenCurrentUser` internally, if succeed then store the returned `HKEY` as a global variable `hUserKey`.

`WinmmRosLogoff` will close the shared `hUserKey` if needed and set it back to the default (predefined HKEY_CURRENT_USER).

Reply:
Or we can just use `RegOverridePredefKey` from the winlogon side?

Reply:
Then since you're going to call RegOpenCurrentUser from winmm, and you're going to use a `hUserKey`, this means that you're going to remove the `RegOpenKeyW(HKEY_CURRENT_USER` call that is currently present, therefore modifying that PlaySound function anyways.

Also, we want (in ReactOS) to *remove* hacks and do actual fixes, not adding new hacks, so this WinmmRosLogon/Logoff ROS-specific exports solution won't be accepted. As mentioned in a past comment, it can be seen that in Vista+ (and Windows 7) there is no such exports anymore, yet sounds work during logon etc.

Reply:
> Can we just use `RegOverridePredefKey` from winlogon side?

And then if you replace Windows' winmm.dll by ours, on Windows, logon sounds and messagebox sounds will stop working correctly?

Reply:
> it can be seen that in Vista+ (and Windows 7) there is no such exports anymore, yet sounds work during logon etc.

Is that requires WASAPI to be implemented?

Reply:
perhaps, but not really. winlogon still uses PlaySound.

Reply:
![Winmm_W7](https://user-images.githubusercontent.com/86486978/188272368-46fca66a-0931-40de-9ca2-9f6ee0aee371.PNG)

So `RegOpenCurrentUser` may be the way Win7 does this thing.

Reply:
> Another HKEY_CURRENT_USER that may be buggy for the same reasons, in winmm, is this one: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/winmm/midimap/midimap.c;hb=83442919e4df0067a7bc72f351fefed667c1737a#l207

Which system process needs this function?

Reply:
**Any** process (can be a user-mode app started by the user) that impersonates a different user (e.g. one that could play the sounds for different users... ?) and calls any of these winmm functions.

Reply:
OK, I'll add a fix for it.